tools/batch-script/address_provider.go

```diff
@@ -0,0 +1,224 @@
+ */
+
+package batch_script
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/client"
+	"github.com/rs/zerolog"
+)
+
+// AddressProvider Is used to get all the addresses that exists at a certain referenceBlockId
+type AddressProvider struct {
+	log              zerolog.Logger
+	lastAddress      flow.Address
+	generator        *flow.AddressGenerator
+	lastAddressIndex uint
+	referenceBlockID flow.Identifier
+	currentIndex     uint
+}
+
+const endOfAccountsError = "storage used is not initialized or not initialized correctly"
+
+const accountStorageUsageScript = `
+pub fun main(address: Address): UInt64 {
+  return getAccount(address).storageUsed
+}
+`
+
+// InitAddressProvider uses bisection to get the last existing address.
+func InitAddressProvider(
+	ctx context.Context,
+	log zerolog.Logger,
+	chain flow.ChainID,
+	referenceBlockID flow.Identifier,
+	client *client.Client,
+	pause time.Duration,
+) (*AddressProvider, error) {
+	ap := &AddressProvider{
+		log:              log,
+		generator:        flow.NewAddressGenerator(chain),
+		referenceBlockID: referenceBlockID,
+		currentIndex:     1,
+	}
+
+	last := time.Now()
+
+	searchStep := 0
+	addressExistsAtIndex := func(index uint) (bool, error) {
+		if time.Since(last) < pause {
+			time.Sleep(pause)
+		}
+		last = time.Now()
+
+		searchStep += 1
+		address := ap.indexToAddress(index)
+
+		// This script will fail with endOfAccountsError
+		// if the account (address at given index) doesn't exist yet
+		_, err := client.ExecuteScriptAtBlockID(
+			ctx,
+			referenceBlockID,
+			[]byte(accountStorageUsageScript),
+			[]cadence.Value{cadence.NewAddress(address)},
+		)
+		if err == nil {
+			return true, nil
+		}
+		if strings.Contains(err.Error(), endOfAccountsError) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	// We assume address #2 exists
+	lastAddressIndex, err := ap.getLastAddress(1, 2, true, addressExistsAtIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info().
+		Str("lastAddress", ap.indexToAddress(lastAddressIndex).Hex()).
+		Uint("numAccounts", lastAddressIndex).
+		Int("stepsNeeded", searchStep).
+		Msg("Found last address")
+
+	ap.lastAddress = ap.indexToAddress(lastAddressIndex)
+	ap.lastAddressIndex = lastAddressIndex
+	return ap, nil
+}
+
+// getLastAddress is a recursive function that finds the last address. Will use max 2 * log2(number_of_addresses) steps
+// If the last address is at index 7 the algorithm goes like this:
+// (x,y) <- lower and upper index
+// 0. start at (1,2); address exists at 2
+// 1. (2,4): address exists at 4
+// 2. (4,8): address doesnt exist at 8
+// 3. (4,8): check address (8 - 4) / 2 = 6  address exists so next pair is (6,8)
+// 4. (6,8): check address 7 address exists so next pair is (7,8)
+// 5. (7,8): check address (8 - 7) / 2 = 7 ... ok already checked so this is the last existing address
+//
+func (p *AddressProvider) getLastAddress(
```

Comment:
Go really needs to get fast generics so this kind of thing can be generalized :/

Reply:
Hmm, how would generics help here, the types are always constant, no?